src/Compiler/Checking/CheckComputationExpressions.fs

```diff
@@ -1399,6 +1402,47 @@ let TcComputationExpression (cenv: cenv) env (overallTy: OverallTy) tpenv (mWhol
             
             Some(translatedCtxt callExpr)
 
+        | SynExpr.WhileBang (spWhile, guardExpr, innerComp, _) -> 
+            let mGuard = guardExpr.Range
+            let mWhile = match spWhile with DebugPointAtWhile.Yes m -> m.NoteSourceConstruct(NotedSourceConstruct.While) | _ -> mGuard
+
+            if isQuery then error(Error(FSComp.SR.tcNoWhileInQuery(), mWhile))
+
+            if isNil (TryFindIntrinsicOrExtensionMethInfo ResultCollectionSettings.AtMostOneResult cenv env mWhile ad "While" builderTy) then
+                error(Error(FSComp.SR.tcRequireBuilderMethod("While"), mWhile))
+
+            if isNil (TryFindIntrinsicOrExtensionMethInfo ResultCollectionSettings.AtMostOneResult cenv env mWhile ad "Delay" builderTy) then
+                error(Error(FSComp.SR.tcRequireBuilderMethod("Delay"), mWhile))
+
+            if isNil (TryFindIntrinsicOrExtensionMethInfo ResultCollectionSettings.AtMostOneResult cenv env mWhile ad "Bind" builderTy) then
+                error(Error(FSComp.SR.tcRequireBuilderMethod("Bind"), mWhile))
+
+            // 'while' is hit just before each time the guard is called
+            let guardExpr = 
+                match spWhile with
+                | DebugPointAtWhile.Yes _ ->
+                    SynExpr.DebugPoint(DebugPointAtLeafExpr.Yes mWhile, false, guardExpr)
+                | DebugPointAtWhile.No -> guardExpr
+
+            // todo desugar directly instead of rewriting first
+            let body =
+                let id = mkSynId range.Zero "$cond"
+                let pat = mkSynPatVar None id
+
+                let body =
+                    let id2 = mkSynId range.Zero "$condM"
+                    let pat2 = mkSynPatVar None id2
+                    let b = mkSynBinding (Xml.PreXmlDoc.Empty, pat2) (None, false, true, range.Zero, DebugPointAtBinding.NoneAtInvisible, None, SynExpr.Ident id, range.Zero, [], [], None, SynBindingTrivia.Zero)  
+                    let set = SynExpr.LongIdentSet (SynLongIdent.SynLongIdent ([ id2 ], [], []), SynExpr.Ident id, range.Zero)
+                    let bang = SynExpr.LetOrUseBang (DebugPointAtBinding.NoneAtInvisible, false, false, pat, guardExpr, [], set, range.Zero, SynExprLetOrUseBangTrivia.Zero)
+
+                    let body = SynExpr.While (spWhile, SynExpr.Ident id2, SynExpr.Sequential (DebugPointAtSequential.SuppressBoth, true, innerComp, bang, range.Zero), range.Zero)
+                    SynExpr.LetOrUse (false, false, [ b ], body, range.Zero, { InKeyword = None })
+
+                SynExpr.LetOrUseBang (DebugPointAtBinding.NoneAtInvisible, false, false, pat, guardExpr, [], body, range.Zero, SynExprLetOrUseBangTrivia.Zero)
+
+            tryTrans CompExprTranslationPass.Initial q varSpace body translatedCtxt
+
```

Comment:
I could not quite wrap my head around translating `while!` directly to builder methods, so for the time being I rewrote the entire expression to a combination of `let!` and `while`, and translated that. The ranges and debug points are obviously all wrong, but the emitted code looks fine so far.

Reply:
On second thought, this might be a valid approach... Please let me know.

Reply:
@T-Gro, the last thing I need help with is this. It works in terms of emitted IL, but there are issues.

All it does is expand

```fsharp
while! cond do
    body
```

to

```fsharp
let! __first = cond
let __cond = first

while __cond do
    body
    let! __first = cond
    __cond <- __first
```

Reply:
I would use a range that will be marked with ".MakeSynthetic()". 
This is a binary mask, and I believe the range.Zero does not have it.. ?

That should help for the tooling scenario at least.
Also, is __cond being marked as mutable somewhere?

Reply:
The updated version already uses synthetic ranges, __cond is indeed mutable.

Reply:
I tried to stare at it for a while, but I do not see a precedence for solving those.
Maybe @KevinRansom  will have an idea?


esp. the "doubled errors" caused when __cond is not a boolean are annoying.

Reply:
Bindings in TAST do have a compiler generated flag, but we're not quite on that level yet.

Reply:
`CompilerGeneratedName` does the trick. `first@` is hidden from Intellisense, and while you can still reach it, it is coupled with a warning.

The doubling of errors remains, though I don't think it's that big of a deal.